### ISSUE TITLE
ice: `Unexpected type for constructor`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#120601]
This triggers only with parallel compiler it seems :thinking: 

snippet:
````rust
struct T;

struct Tuple(i32);

async fn foo() -> Result<(), ()> {
    Unstable2(())
}

async fn tuple() -> Tuple {
    Tuple(1i32)
}

async fn match_() {
    match tuple() {
        Tuple(_) => {}
    }
}


````

Version information
````
rustc 1.77.0-nightly (bf3c6c5be 2024-02-01)
binary: rustc
commit-hash: bf3c6c5bed498f41ad815641319a1ad9bcecb8e8
commit-date: 2024-02-01
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zthreads=40`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0670]: `async fn` is not permitted in Rust 2015
 --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:5:1
  |
5 | async fn foo() -> Result<(), ()> {
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2021` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0670]: `async fn` is not permitted in Rust 2015
 --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:9:1
  |
9 | async fn tuple() -> Tuple {
  | ^^^^^ to use `async fn`, switch to Rust 2018 or later
  |
  = help: pass `--edition 2021` to `rustc`
  = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0670]: `async fn` is not permitted in Rust 2015
  --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:13:1
   |
13 | async fn match_() {
   | ^^^^^ to use `async fn`, switch to Rust 2018 or later
   |
   = help: pass `--edition 2021` to `rustc`
   = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:17:2
   |
17 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs`

error[E0308]: mismatched types
  --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:15:9
   |
14 |     match tuple() {
   |           ------- this expression has type `impl Future<Output = Tuple>`
15 |         Tuple(_) => {}
   |         ^^^^^^^^ expected future, found `Tuple`
   |
   = note: expected opaque type `impl Future<Output = Tuple>`
                   found struct `Tuple`
help: consider `await`ing on the `Future`
   |
14 |     match tuple().await {
   |                  ++++++

error: internal compiler error: compiler/rustc_pattern_analysis/src/rustc.rs:240:22: Unexpected type for constructor `Struct`: Alias(Opaque, AliasTy { args: [], def_id: DefId(0:15 ~ mvce[954a]::tuple::{opaque#0}) })

thread 'rustc' panicked at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/compiler/rustc_errors/src/lib.rs:932:30:
Box<dyn Any>
stack backtrace:
error[E0425]: cannot find function, tuple struct or tuple variant `Unstable2` in this scope
 --> /tmp/icemaker_global_tempdir.DSSWU6jJMtCk/rustc_testrunner_tmpdir_reporting.TVo0XDALOMJR/mvce.rs:6:5
  |
6 |     Unstable2(())
  |     ^^^^^^^^^ not found in this scope

   0:     0x7f0cebb8bd16 - std::backtrace_rs::backtrace::libunwind::trace::he544685250360e22
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f0cebb8bd16 - std::backtrace_rs::backtrace::trace_unsynchronized::h664adef12633380a
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f0cebb8bd16 - std::sys_common::backtrace::_print_fmt::hba7ee045aeee07b0
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f0cebb8bd16 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc80b7ff177a688f2
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f0cebbde770 - core::fmt::rt::Argument::fmt::h605309ab4829b6b5
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/core/src/fmt/rt.rs:142:9
   5:     0x7f0cebbde770 - core::fmt::write::h5f24f526de99e105
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f0cebb7f61f - std::io::Write::write_fmt::h711c920f97cd3d8c
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/io/mod.rs:1846:15
   7:     0x7f0cebb8baf4 - std::sys_common::backtrace::_print::ha279702af2871260
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f0cebb8baf4 - std::sys_common::backtrace::print::hc1f3820f405abdd3
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f0cebb8e887 - std::panicking::default_hook::{{closure}}::h5058841d19c78172
  10:     0x7f0cebb8e5e9 - std::panicking::default_hook::h60a42497ec033f97
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/panicking.rs:292:9
  11:     0x7f0cee9371fc - std[4337b1e3d1c66af6]::panicking::update_hook::<alloc[6a28242b4f343ab9]::boxed::Box<rustc_driver_impl[e51f6499da6a94b9]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f0cebb8efd6 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h437fd8fcb0ff0cea
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/alloc/src/boxed.rs:2029:9
  13:     0x7f0cebb8efd6 - std::panicking::rust_panic_with_hook::h294fbe090fe44fa4
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/panicking.rs:785:13
  14:     0x7f0cee9693b4 - std[4337b1e3d1c66af6]::panicking::begin_panic::<rustc_errors[37e82e04e08bc148]::ExplicitBug>::{closure#0}
  15:     0x7f0cee965aa6 - std[4337b1e3d1c66af6]::sys_common::backtrace::__rust_end_short_backtrace::<std[4337b1e3d1c66af6]::panicking::begin_panic<rustc_errors[37e82e04e08bc148]::ExplicitBug>::{closure#0}, !>
  16:     0x7f0cee960d06 - std[4337b1e3d1c66af6]::panicking::begin_panic::<rustc_errors[37e82e04e08bc148]::ExplicitBug>
  17:     0x7f0cee974591 - <rustc_errors[37e82e04e08bc148]::diagnostic_builder::BugAbort as rustc_errors[37e82e04e08bc148]::diagnostic_builder::EmissionGuarantee>::emit_producing_guarantee
  18:     0x7f0ceed2cbcb - <rustc_errors[37e82e04e08bc148]::DiagCtxt>::bug::<alloc[6a28242b4f343ab9]::string::String>
  19:     0x7f0ceedc3ceb - rustc_middle[9b442554c155189d]::util::bug::opt_span_bug_fmt::<rustc_span[ecf473d7733698d2]::span_encoding::Span>::{closure#0}
  20:     0x7f0ceedaca2a - rustc_middle[9b442554c155189d]::ty::context::tls::with_opt::<rustc_middle[9b442554c155189d]::util::bug::opt_span_bug_fmt<rustc_span[ecf473d7733698d2]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f0ceedac8c8 - rustc_middle[9b442554c155189d]::ty::context::tls::with_context_opt::<rustc_middle[9b442554c155189d]::ty::context::tls::with_opt<rustc_middle[9b442554c155189d]::util::bug::opt_span_bug_fmt<rustc_span[ecf473d7733698d2]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f0cecd61d40 - rustc_middle[9b442554c155189d]::util::bug::bug_fmt
  23:     0x7f0cf058517f - rustc_pattern_analysis[194e970a029e56db]::usefulness::compute_exhaustiveness_and_usefulness::<rustc_pattern_analysis[194e970a029e56db]::rustc::RustcMatchCheckCtxt>::{closure#0}
  24:     0x7f0cf0579617 - rustc_pattern_analysis[194e970a029e56db]::analyze_match
  25:     0x7f0cecbcc766 - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  26:     0x7f0cecbcbb8d - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  27:     0x7f0cecbccac3 - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  28:     0x7f0cecbcbb8d - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  29:     0x7f0cecbccac3 - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  30:     0x7f0cecbcbb8d - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  31:     0x7f0cecbccac3 - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  32:     0x7f0cecbcbb8d - <rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::MatchVisitor as rustc_middle[9b442554c155189d]::thir::visit::Visitor>::visit_expr
  33:     0x7f0ceff723fa - rustc_mir_build[255f66b20b49daa2]::thir::pattern::check_match::check_match
  34:     0x7f0ceff7208f - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::check_match::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>
  35:     0x7f0ceff7161e - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  36:     0x7f0ceff712c3 - rustc_query_impl[5a78334436849816]::query_impl::check_match::get_query_non_incr::__rust_end_short_backtrace
  37:     0x7f0ceff8144a - rustc_mir_build[255f66b20b49daa2]::build::mir_built
  38:     0x7f0ceff81315 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  39:     0x7f0cefd88d3a - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  40:     0x7f0cefd88850 - rustc_query_impl[5a78334436849816]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7f0cefd02fb9 - rustc_mir_transform[a2d391ebd38b778]::ffi_unwind_calls::has_ffi_unwind_calls
  42:     0x7f0cefd02baf - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::has_ffi_unwind_calls::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>
  43:     0x7f0ceff71553 - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  44:     0x7f0ceff7121d - rustc_query_impl[5a78334436849816]::query_impl::has_ffi_unwind_calls::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7f0ceff5e60f - rustc_mir_transform[a2d391ebd38b778]::mir_const
  46:     0x7f0ceff5e549 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::mir_const::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  47:     0x7f0cefd88d3a - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  48:     0x7f0cefd88790 - rustc_query_impl[5a78334436849816]::query_impl::mir_const::get_query_non_incr::__rust_end_short_backtrace
  49:     0x7f0cede3f952 - rustc_mir_transform[a2d391ebd38b778]::mir_promoted
  50:     0x7f0cf0418c12 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 16usize]>>
  51:     0x7f0cf0418ec6 - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  52:     0x7f0cf0a18055 - rustc_query_impl[5a78334436849816]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  53:     0x7f0cf0a1816e - rustc_borrowck[53d443e82d94d72b]::mir_borrowck
  54:     0x7f0cf0a18099 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  55:     0x7f0cefd88d3a - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  56:     0x7f0cefd886d0 - rustc_query_impl[5a78334436849816]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  57:     0x7f0cf00c7149 - <rustc_borrowck[53d443e82d94d72b]::type_check::TypeChecker>::prove_closure_bounds
  58:     0x7f0cf09e7ef3 - <rustc_borrowck[53d443e82d94d72b]::type_check::TypeChecker>::typeck_mir
  59:     0x7f0ced216342 - rustc_borrowck[53d443e82d94d72b]::type_check::type_check
  60:     0x7f0ced1fb3be - rustc_borrowck[53d443e82d94d72b]::nll::compute_regions
  61:     0x7f0cf0a2830f - rustc_borrowck[53d443e82d94d72b]::do_mir_borrowck
  62:     0x7f0cf0a18659 - rustc_borrowck[53d443e82d94d72b]::mir_borrowck
  63:     0x7f0cf0a18099 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  64:     0x7f0cefd88d3a - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  65:     0x7f0cefd886d0 - rustc_query_impl[5a78334436849816]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  66:     0x7f0cf0415a22 - rustc_middle[9b442554c155189d]::query::plumbing::query_get_at::<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_span[ecf473d7733698d2]::def_id::LocalDefId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>>
  67:     0x7f0cf0bd3b62 - rustc_hir_analysis[1563c848cf98af67]::collect::type_of::type_of_opaque
  68:     0x7f0cf0bd39c9 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::type_of_opaque::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  69:     0x7f0cefc75361 - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::DefIdCache<rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  70:     0x7f0cf0c2cd1e - rustc_query_impl[5a78334436849816]::query_impl::type_of_opaque::get_query_non_incr::__rust_end_short_backtrace
  71:     0x7f0cf015a6a3 - rustc_middle[9b442554c155189d]::query::plumbing::query_get_at::<rustc_query_system[f1aaee53077b1b15]::query::caches::DefIdCache<rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>>
  72:     0x7f0cede9a0f0 - rustc_hir_analysis[1563c848cf98af67]::collect::type_of::type_of
  73:     0x7f0cefc766ea - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>
  74:     0x7f0cefc75361 - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::DefIdCache<rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  75:     0x7f0cefc74f1d - rustc_query_impl[5a78334436849816]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  76:     0x7f0cf015a6a3 - rustc_middle[9b442554c155189d]::query::plumbing::query_get_at::<rustc_query_system[f1aaee53077b1b15]::query::caches::DefIdCache<rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 8usize]>>>
  77:     0x7f0cf065b99f - rustc_hir_analysis[1563c848cf98af67]::check::wfcheck::check_well_formed
  78:     0x7f0cf0658d59 - rustc_query_impl[5a78334436849816]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[5a78334436849816]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>
  79:     0x7f0cf0658431 - rustc_query_system[f1aaee53077b1b15]::query::plumbing::try_execute_query::<rustc_query_impl[5a78334436849816]::DynamicConfig<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_hir[1f4f973af63e82cc]::hir_id::OwnerId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[5a78334436849816]::plumbing::QueryCtxt, false>
  80:     0x7f0cf0658191 - rustc_query_impl[5a78334436849816]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  81:     0x7f0cf0655e53 - rustc_middle[9b442554c155189d]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[f1aaee53077b1b15]::query::caches::VecCache<rustc_hir[1f4f973af63e82cc]::hir_id::OwnerId, rustc_middle[9b442554c155189d]::query::erase::Erased<[u8; 1usize]>>, ()>
  82:     0x7f0ceea01d72 - rayon[e3841394ac71e4b3]::iter::plumbing::bridge_producer_consumer::helper::<rayon[e3841394ac71e4b3]::slice::IterProducer<rustc_hir[1f4f973af63e82cc]::hir::ItemId>, rayon[e3841394ac71e4b3]::iter::filter_map::FilterMapConsumer<rayon[e3841394ac71e4b3]::iter::reduce::ReduceConsumer<<core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>>::and<()>, rustc_data_structures[396196f1dd43bbd8]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[1f4f973af63e82cc]::hir::ItemId], rustc_span[ecf473d7733698d2]::ErrorGuaranteed, <rustc_middle[9b442554c155189d]::hir::ModuleItems>::par_items<rustc_hir_analysis[1563c848cf98af67]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}>, rustc_data_structures[396196f1dd43bbd8]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[1f4f973af63e82cc]::hir::ItemId], rustc_span[ecf473d7733698d2]::ErrorGuaranteed, <rustc_middle[9b442554c155189d]::hir::ModuleItems>::par_items<rustc_hir_analysis[1563c848cf98af67]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}>>
  83:     0x7f0ceea24793 - <rayon_core[78331a983d6f9d5d]::job::StackJob<rayon_core[78331a983d6f9d5d]::latch::SpinLatch, rayon_core[78331a983d6f9d5d]::join::join_context::call_b<core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>, rayon[e3841394ac71e4b3]::iter::plumbing::bridge_producer_consumer::helper<rayon[e3841394ac71e4b3]::slice::IterProducer<rustc_hir[1f4f973af63e82cc]::hir::ItemId>, rayon[e3841394ac71e4b3]::iter::filter_map::FilterMapConsumer<rayon[e3841394ac71e4b3]::iter::reduce::ReduceConsumer<<core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>>::and<()>, rustc_data_structures[396196f1dd43bbd8]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[1f4f973af63e82cc]::hir::ItemId], rustc_span[ecf473d7733698d2]::ErrorGuaranteed, <rustc_middle[9b442554c155189d]::hir::ModuleItems>::par_items<rustc_hir_analysis[1563c848cf98af67]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#1}>, rustc_data_structures[396196f1dd43bbd8]::sync::parallel::enabled::try_par_for_each_in<&[rustc_hir[1f4f973af63e82cc]::hir::ItemId], rustc_span[ecf473d7733698d2]::ErrorGuaranteed, <rustc_middle[9b442554c155189d]::hir::ModuleItems>::par_items<rustc_hir_analysis[1563c848cf98af67]::check::wfcheck::check_mod_type_wf::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}>>::{closure#1}>::{closure#0}, core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>> as rayon_core[78331a983d6f9d5d]::job::Job>::execute
  84:     0x7f0cee53b915 - <rayon_core[78331a983d6f9d5d]::registry::WorkerThread>::wait_until_cold
  85:     0x7f0cee538403 - <rayon_core[78331a983d6f9d5d]::registry::ThreadBuilder>::run
  86:     0x7f0cee939437 - <<crossbeam_utils[e79582213e43f505]::thread::ScopedThreadBuilder>::spawn<<rayon_core[78331a983d6f9d5d]::ThreadPoolBuilder>::build_scoped<rustc_interface[e57535df52585ed5]::util::run_in_thread_pool_with_globals<rustc_interface[e57535df52585ed5]::interface::run_compiler<core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>, rustc_driver_impl[e51f6499da6a94b9]::run_compiler::{closure#0}>::{closure#0}, core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>>::{closure#3}::{closure#0}::{closure#0}, rustc_interface[e57535df52585ed5]::util::run_in_thread_pool_with_globals<rustc_interface[e57535df52585ed5]::interface::run_compiler<core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>, rustc_driver_impl[e51f6499da6a94b9]::run_compiler::{closure#0}>::{closure#0}, core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>>::{closure#3}::{closure#0}::{closure#1}, core[cc6cd760d986f799]::result::Result<(), rustc_span[ecf473d7733698d2]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, ()>::{closure#0} as core[cc6cd760d986f799]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  87:     0x7f0cee92ebce - std[4337b1e3d1c66af6]::sys_common::backtrace::__rust_begin_short_backtrace::<alloc[6a28242b4f343ab9]::boxed::Box<dyn core[cc6cd760d986f799]::ops::function::FnOnce<(), Output = ()> + core[cc6cd760d986f799]::marker::Send>, ()>
  88:     0x7f0cee9390d3 - <<std[4337b1e3d1c66af6]::thread::Builder>::spawn_unchecked_<alloc[6a28242b4f343ab9]::boxed::Box<dyn core[cc6cd760d986f799]::ops::function::FnOnce<(), Output = ()> + core[cc6cd760d986f799]::marker::Send>, ()>::{closure#1} as core[cc6cd760d986f799]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  89:     0x7f0cebb98675 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc4b9a6efba849e51
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/alloc/src/boxed.rs:2015:9
  90:     0x7f0cebb98675 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hc525ee5ac423ddcd
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/alloc/src/boxed.rs:2015:9
  91:     0x7f0cebb98675 - std::sys::pal::unix::thread::Thread::new::thread_start::h31b81f67226864c0
                               at /rustc/bf3c6c5bed498f41ad815641319a1ad9bcecb8e8/library/std/src/sys/pal/unix/thread.rs:108:17
  92:     0x7f0ceb9839eb - <unknown>
  93:     0x7f0ceba077cc - <unknown>
  94:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.77.0-nightly (bf3c6c5be 2024-02-01) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=40 -Z dump-mir-dir=dir

query stack during panic:
#0 [check_match] match-checking `match_::{closure#0}`
#1 [mir_built] building MIR for `match_::{closure#0}`
#2 [has_ffi_unwind_calls] checking if `match_::{closure#0}` contains FFI-unwind calls
#3 [mir_const] preparing `match_::{closure#0}` for borrow checking
#4 [mir_promoted] promoting constants in MIR for `match_::{closure#0}`
#5 [mir_borrowck] borrow-checking `match_::{closure#0}`
#6 [mir_borrowck] borrow-checking `match_`
#7 [type_of_opaque] computing type of opaque `match_::{opaque#0}`
#8 [type_of] computing type of `match_::{opaque#0}`
#9 [check_well_formed] checking that `match_::{opaque#0}` is well-formed
#10 [check_mod_type_wf] checking that types are well-formed in top-level module
#11 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 7 previous errors

Some errors have detailed explanations: E0308, E0425, E0601, E0670.
For more information about an error, try `rustc --explain E0308`.

```

</p>
</details>

**Comment by matthiaskrgr** [comment]
Ouff this bisects to #117213 cc @oli-obk

**Comment by oli-obk** [comment]
That PR just exposed the issue. It was preexisting

**Comment by Zoxc** [comment]
I'm not able to reproduce this with `496145b9cc023aef4bb1f16c0964a53d0da36c88`. Perhaps it was fixed by 
https://github.com/rust-lang/rust/pull/127311?
